Task_18

evTyurin opened the pull request.

YevganyTyurin committed:
Add tests for CraftsmanService and GaragePlaceService

mrseneca commented on carservice/src/test/java/com/senlainc/warsaw/tyurin/CraftsmanServiceUnitTest.java:
```diff
@@ -0,0 +1,177 @@
+    @Test
+    void find_findAllCraftsmen_returnedCraftsmen() {
+        Craftsman testCraftsman1 = new Craftsman();
+        testCraftsman1.setId(1L);
```
DRY, if you using same test data in each tests better create consts with this data and use it, that's will help if you need to change data, in that way it only require once. No need to repeat same code in every test

mrseneca commented on carservice/src/test/java/com/senlainc/warsaw/tyurin/CraftsmanServiceUnitTest.java:
```diff
@@ -0,0 +1,177 @@
+        testCraftsman.setName("John");
+        testCraftsman.setSurname("Constantine");
+
+        when(craftsmanDao.findById(1L)).thenReturn(testCraftsman);
```
Same, create const ID and pass it in every test, if you decide to change it, it will be in only one place

mrseneca commented on carservice/src/test/java/com/senlainc/warsaw/tyurin/GaragePlaceServiceUnitTest.java:
```diff
@@ -0,0 +1,117 @@
+
+    @Test
+    void find_findAvailableGaragePlacesAmount_returnedAmount() {
+        when(garagePlaceDao.getAvailablePlacesAmount(LocalDateTime
```
Same here, code duplication, use const or variable

mrseneca commented on carservice/src/test/java/com/senlainc/warsaw/tyurin/CraftsmanServiceUnitTest.java:
```diff
@@ -0,0 +1,177 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class CraftsmanServiceUnitTest {
```
Cant find create craftsman test

mrseneca replied:
Can't find any tests for getSortedByCriteria, it's perfect for unit testing

mrseneca commented on carservice/src/test/java/com/senlainc/warsaw/tyurin/GaragePlaceServiceUnitTest.java:
```diff
@@ -0,0 +1,117 @@
+        testGaragePlace.setNumber(1);
+        testGaragePlace.setSpace(15.0);
+
+        ReflectionTestUtils.setField(garagePlaceService,"isGaragePlaceRemovable",true);
```
Using ReflectionUtils in tests is generally discouraged due to several reasons. Reflection in Java provides dynamic access to fields, methods, and constructors during runtime, but it comes with trade-offs that can impact code readability, maintainability, and robustness. Here are some key reasons why it's not recommended:
   - Code that relies heavily on reflection can be difficult to read and understand, making it harder for developers to grasp the logic and intentions behind the tests.
   - Reflection bypasses the usual compile-time checks, making the code more prone to errors. If the structure of the classes being reflected upon changes, it can break the tests, and these issues might not be caught until runtime.
   - Many testing tools and frameworks are designed to work seamlessly with standard testing practices. Reflection-based code may not be as compatible with these tools, leading to a lack of support for automated testing and other testing utilities.
   - Debugging reflective code can be more challenging than debugging regular code. Developers may find it harder to trace and understand the flow of the program during debugging sessions.
   - Reflection can have performance overhead, and using it excessively in tests may impact the overall test suite performance. Tests should ideally run quickly and efficiently.
   - The need for extensive use of reflection in tests might indicate design issues in the codebase. It's often better to address these design concerns rather than resorting to reflection for testing purposes.

mrseneca replied:
To fix issue with value you can use test resources, with desired data

@TestPropertySource(locations = "classpath:test-application.properties")

evTyurin replied:
Thank you for detailed explanation
@TestPropertySource(locations = "classpath:test-application.properties") is what I need in this situation

mrseneca reviewed (changes requested):
Well done, fix code duplication and also check if all methods are covered by tests, I found few in a first test class may be I miss something in second.

mrseneca commented:
Also check your repo issues, I created one for you to fix. About user password saving

YevganyTyurin committed:
Update tests for CraftsmanService and GaragePlaceService.
Add tests for CustomUserDetailsService, OrderService, AuthService.
Add password encryption

mrseneca reviewed (approved):
Well done
